### PR TITLE
truncate i8-s to i1-s when loading constants

### LINKED ISSUE #30891: LLVM Assertion Error
Getting an LLVM assertion error on nightly inside of a macro. The error doesn't occur on stable.

The error:

> Assertion failed: (C1->getType() == C2->getType() && "Operand types in binary constant expression should match"), function get, file /Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/src/llvm/lib/IR/Constants.cpp, line 1875.

```rust
const ERROR_CONST: bool = false;

macro_rules! cause_llvm_assertion_error {
    ($a: expr, $b: expr, $c: expr) => (
        if $a || $b || !($c) {
            println!("assertion error?");
        }
    )
}

pub fn main() {
    cause_llvm_assertion_error!(true, false, ERROR_CONST);
}
```

### PR DESCRIPTION
**Pull request by arielb1**
Fixes #30891

r? @eddyb

### CODE REVIEW

File: src/librustc_trans/trans/consts.rs [hunk 1]
```diff
@@ -161,13 +161,17 @@ pub fn addr_of(ccx: &CrateContext,
...
+fn load_const(cx: &CrateContext, v: ValueRef, t: Ty) -> ValueRef {
     let v = match cx.const_unsized().borrow().get(&v) {
         Some(&v) => v,
         None => v
     };
-    unsafe {
-        llvm::LLVMGetInitializer(v)
+    let d = unsafe { llvm::LLVMGetInitializer(v) };
+    if t.is_bool() {
+        unsafe { llvm::LLVMConstTrunc(d, Type::i1(cx).to_ref()) }
```
Comment (eddyb) [inline src/librustc_trans/trans/consts.rs]: Isn't there something similar in `base::load_ty`, using `LLVMGetInitializer`?

File: src/test/run-pass/issue-30891.rs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+const ERROR_CONST: bool = true;
```
Comment (eddyb) [inline src/test/run-pass/issue-30891.rs]: Does this work without a license header?

Comment (arielb1) [inline src/test/run-pass/issue-30891.rs]: Fixing that.

File: src/librustc_trans/trans/consts.rs (hunk 1 above)
Comment (arielb1) [inline src/librustc_trans/trans/consts.rs]: This is a variant of `load_ty` that uses the const llvm functions.

Comment (eddyb) [inline src/librustc_trans/trans/consts.rs]: Nevermind, `load_ty` already handles this correctly via `to_arg_ty`.

**Comment by arielb1** [comment]
nominating for beta since this fixes a regression.

**Pull request merged**